Fix combo marker block restriction issues

GiudGiud opened the pull request:
refs #30103

GiudGiud committed:
Add support for block restriction to combo marker
closes #30103

GiudGiud committed:
Add a test for block restriction support to combo marker

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2750563/), step Docs: sync website on 02f8a2e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30104/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2750579/), step Generate coverage on 02f8a2e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3403d2</th>
      <th colspan=4>#30104 02f8a2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.29%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>109051</td>
      <td>109068</td>
      <td>+17</td>
      <td>18</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18808</td>
      <td>18806</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30104/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30104/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30104/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on framework/include/markers/ComboMarker.h:
```diff
@@ -24,7 +24,15 @@ class ComboMarker : public Marker
 
-  std::vector<MarkerName> _names;
+  /// Names of the markers contributing to the combo
+  const std::vector<MarkerName> _names;
```
```suggestion
  const std::vector<MarkerName> & _names;
```

lindsayad commented on framework/src/markers/ComboMarker.C:
```diff
@@ -24,10 +25,31 @@ ComboMarker::validParams()
 ComboMarker::ComboMarker(const InputParameters & parameters)
-  : Marker(parameters), _names(parameters.get<std::vector<MarkerName>>("markers"))
+  : Marker(parameters),
+    _names(parameters.get<std::vector<MarkerName>>("markers")),
```
```suggestion
    _names(getParam<std::vector<MarkerName>>("markers")),
```

lindsayad commented on framework/src/markers/ComboMarker.C:
```diff
@@ -24,10 +25,31 @@ ComboMarker::validParams()
   for (const auto & marker_name : _names)
     _markers.push_back(&getMarkerValue(marker_name));
+
+  for (const auto & marker_name : _names)
+    _marker_variables.push_back(&_fe_problem.getVariable(_tid, marker_name));
+
+  // Check block restrictions
+  std::string other_block_restricted = "";
+  for (const auto & marker_name : _names)
+    if (blockIDs() != _fe_problem.getVariable(_tid, marker_name).blockIDs())
```
Can you consolidate the two loops, only call `_subproblem.getVariable` once?

GiudGiud replied:
done

lindsayad commented on framework/src/markers/ComboMarker.C:
```diff
@@ -24,10 +25,31 @@ ComboMarker::validParams()
     _markers.push_back(&getMarkerValue(marker_name));
+
+  for (const auto & marker_name : _names)
+    _marker_variables.push_back(&_fe_problem.getVariable(_tid, marker_name));
```
```suggestion
    _marker_variables.push_back(&_subproblem.getVariable(_tid, marker_name));
```

lindsayad replied:
Markers may be placed on the displaced mesh vs. undisplaced

lindsayad commented on test/tests/markers/combo_marker/tests:
```diff
@@ -1,13 +1,19 @@
+    type = 'Exodiff'
+    input = 'combo_marker_block_restricted.i'
+    exodiff = 'combo_marker_block_restricted_out.e'
+    requirement = "The system shall include the ability to combine multiple mesh refinement markers into a single value even if the block restriction of each marker do not match."
```
```suggestion
    requirement = "The system shall include the ability to combine multiple mesh refinement markers into a single value even if the block restrictions of each marker do not match."
```

GiudGiud committed:
Address Alex's review

get markers from subproblem in case they are on displaced mesh
use a reference for vector of names

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>
